### ISSUE TITLE
Workflow processing has been postponed because too many workflows are already running

### ISSUE THREAD

**Issue by tooptoop4** [argoproj/argo-workflows#11984]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

i have parallelism set to 30 in configmap of workflow controller but even with something like 17 workflows running (confirmed from ui and cli) it can sometimes keep workflows in pending state with message 'Workflow processing has been postponed because too many workflows are already running' for a long time

### Version

3.4.11

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
n/a
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

**Comment by tooptoop4** [comment]
they only rely on parallelism but there were other workflows waiting on a mutex:
```
NAME                                               STATUS      AGE    MESSAGE
aaaa-dkntk                                         Running     32m
redact-20231010t044555074ztkhzp                    Running     11m
redact-20231010t044558032z9q6xp                    Pending            Waiting for auth/Mutex/tbl0/ lock. Lock status: 0/1
redact-20231010t044602469zh999j                    Pending            Waiting for auth/Mutex/tbl0/ lock. Lock status: 0/1
redact-20231010t044608533zhn6g4                    Pending            Waiting for auth/Mutex/tbl0/ lock. Lock status: 0/1
redact-20231010t044608554ztjsfn                    Pending            Waiting for auth/Mutex/tbl0/ lock. Lock status: 0/1
redact-20231010t044614762ztqslg                    Pending            Waiting for auth/Mutex/tbl0/ lock. Lock status: 0/1
redact-20231010t044615441zxngwl                    Pending            Waiting for auth/Mutex/tbl0/ lock. Lock status: 0/1
redact-20231010t044620131zrsghg                    Pending            Waiting for auth/Mutex/tbl0/ lock. Lock status: 0/1
redact-20231010t044625096z5mjlm                    Pending            Waiting for auth/Mutex/tbl0/ lock. Lock status: 0/1
redact-20231010t050207392z59sx2                    Running     18m
redact-20231010t051126055zh95pj                    Succeeded   81s
redact-20231010t051129543zd9hnz                    Succeeded   60s
redact-20231010t051132641z8ms95                    Succeeded   39s
redact-20231010t051135650z8zm7t                    Running     18s
redact-20231010t051138652zxxrdr                    Pending            Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1
redact-20231010t051141719zscr8w                    Pending            Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1
redact-20231010t051144681zgq85k                    Pending            Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1
redact-20231010t051147586zj5jzg                    Pending            Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1
redact-20231010t051150125zgj4z8                    Pending            Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1
redact-20231010t051154674zbzg8s                    Succeeded   105s
redact-20231010t051154701zx9pxp                    Pending            Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1
redact-20231010t051155670zgb28v                    Succeeded   84s
redact-20231010t051156610zm9tbm                    Succeeded   72s
redact-20231010t051156641zkkz8b                    Succeeded   61s
redact-20231010t051157606zg8q2m                    Pending            Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1
redact-20231010t051158805zlrn7b                    Succeeded   56s
redact-20231010t051159638zqxxsz                    Succeeded   51s
redact-20231010t051200727zxvvfw                    Pending            Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1
redact-20231010t051202173zks6wg                    Running     30s
redact-20231010t051202599zv48rk                    Succeeded   30s
redact-20231010t051203573z6mbvh                    Running     21s
redact-20231010t051203646z95d8p                    Running     16s
redact-20231010t051205828z27nbg                    Running     9s
redact-20231010t051205860zp7klm                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051206749zrfpzh                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051208238zz225v                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051208679z6hp8c                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051209090zfbxx2                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051210187zkjrzb                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051211623zhzn5h                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051211624zwtzzq                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051212832zxt6hk                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051213787zwj5h2                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051213830zxmt55                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051215259zhthhs                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051216465zbx98x                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051217431zjgmwb                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051217443zpmfh2                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051217613z7zbld                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051218333zxlhw2                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051219498zlkrpl                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051219738zhf2p2                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051220765zbpthx                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051221919zr8wf6                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051223310zvjvtr                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051225280zm9tdb                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051226339zlvx5q                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051227360z4n4ld                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051228594z8p89n                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051230186zkswhp                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051231216zshb2q                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051232257zz45d9                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051233599zcjlkt                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051506502zxvjjc                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051506508zn89kr                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051506523zfgrfk                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507351zdbjj7                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507356z6z42r                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507366zk4ztv                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507368zrlxbr                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507373zjfb26                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507380zc49rg                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507390zsqg8z                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507405znhwlc                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051507410zj4dh8                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051511895z48dc9                    Pending            Workflow processing has been postponed because too many workflows are already running
redact-20231010t051511914zhgrfn                    Pending            Workflow processing has been postponed because too many workflows are already running
diffwf5-0ldszc                                     Pending            Workflow processing has been postponed because too many workflows are already running
diffwf5-0rlkkt                                     Pending            Workflow processing has been postponed because too many workflows are already running
diffwf5-04f5jm                                     Pending            Workflow processing has been postponed because too many workflows are already running
diffwf1-oneoff-glsnf                               Running     66m
diffwf2-pftqf                                      Running     70m
diffwf3-1696914900                                 Running     21s
diffwf4-1696905000                                 Running     172m

```

**Comment by agilgur5** [comment]
hmm -- does the sum total of those running plus the ones waiting on a mutex equal to the total? maybe it's counting those incorrectly.
In your current list there's 12 Running + 16 Pending = 28 total. Still under 30 but it seems like some just started so the other 2 may start in the next queue cycle.

If you're able to confirm that then I think we've found a bug. If not, still some brainstorming to do.

**Comment by sarabala1979** [comment]
@tooptoop4  `parallelism ` and ` mutex` are handled differently in the Workflow controller. `parallelism` is top-level to make sure the controller will execute a configured number of workflow concurrently (it will be rate-limited before starting executing workflow). But mutex is happening after the fact (The controller is already considering those workflows are start processing by the controller). This is expected behavior by the controller.

Is there any reason you want use `parallelism` and `mutex`?

**Comment by agilgur5** [comment]
> Is there any reason you want use `parallelism` and `mutex`?

@sarabala1979 there are many reasons I can think of for that; they serve separate purposes.
For instance, you have a `mutex` for Workflows to not use the same table simultaneously (as @tooptoop4 has), while you also want to limit the load of Workflows on your cluster with a `parallelism` cap.

**Comment by sarabala1979** [comment]
>  In your current list there's 12 Running + 16 Pending on mutexes = 28 total. 

As mentioned above it is expected behavior  with Parallelism+ mutex

>  it can sometimes keep workflows in pending state with message 'Workflow processing has been postponed because too many workflows are already running' for a long time

@tooptoop4  Are you seeing `running` or `Waiting for auth/Mutex/tbl1/ lock. Lock status: 0/1` workflows total =30 during the long time?

**Comment by agilgur5** [comment]
> As mentioned above it is expected behavior with Parallelism+ mutex

That may be expected behavior if you are familiar with the code, but I think @tooptoop4 has a point here that a reasonable user would be surprised by this behavior.
The Workflows are not running, yet they are calculated as if they are -- that is surprising.

**Comment by sarabala1979** [comment]
I am not talking about code. As per the feature Parallism+ mutex, it is expected.

As long as, As you mentioned above `12Running + 16 Pending on mutexes= 30Approx` is equal to configured parallelism and keeps moving workflow when current workflows are completed. feature works 

If workflows are not moving even current mutex hold workflow completed, Then there is bug

**Comment by agilgur5** [comment]
> I am not talking about code. As per the feature Parallism+ mutex, it is expected.

This is not documented to work this way though: https://argoproj.github.io/argo-workflows/synchronization/

> feature works
> Then there is bug

Sure we can say that this is working without bugs and call this a feature request instead. But the working behavior is nonetheless surprising

**Comment by sarabala1979** [comment]
> This is not documented to work this way though: https://argoproj.github.io/argo-workflows/synchronization/
Agreed. This scenario needs to be updated in the document. 

> Sure we can say that this is working without bugs and call this a feature request instead. But the working behavior is nonetheless surprising
Agreed.  Thanks @agilgur5 

@tooptoop4 Feature requests and document improvement issues can created.

**Comment by tooptoop4** [comment]
i don't mind if u tag this as 'feature request', am really keen for this functionality!

**Comment by tooptoop4** [comment]
rough guess is https://github.com/argoproj/argo-workflows/blob/v3.5.4/workflow/controller/operator.go#L257-L265 should also be checked at https://github.com/argoproj/argo-workflows/blob/v3.5.4/workflow/controller/controller.go#L776

**Comment by mugdha-adhav** [comment]
Any progress/update on this?

Also we confirmed that this impacts `namespace-parallelism` limit as well.

**Comment by isubasinghe** [comment]
This is not a bug in my eyes and is very much expected behaviour. I suppose if there is some confusion here, we could change the docs such that this behaviour is more clear.

**Comment by mugdha-adhav** [comment]
> we could change the docs such that this behaviour is more clear

Yes please.

The confusion arises because -
- When we set parallelism (or namespace-parallelism) limit, users generally consider running workflows count should be equivalent to limit set, instead of considering workflows being processed.
- Secondly, if semaphore workflows restrict the no. of running workflows because they're waiting on lock, users generally expect that non-semaphore based workflows should run, if total running workflows count hasn't reached parallelism limit.
